blog/2022/storage-map.adoc

```diff
@@ -0,0 +1,127 @@
+
+- In Keycloak 20, expect improved support especially for CockroachDB;
+  and also a file-based store. Together with another anticipated feature
+  - _tree_ store - it would be possible to combine several storage mechanisms
+  and have e.g. several static client declarations in static files managed
+  in a versioning system like Git, combined with dynamic clients stored in a database.
+  Per-request object caching is on radar for Keycloak 20 too and should result in
+  significantly better performance.
+
+- In Keycloak 21, expect LDAP support, and an offline tool for migrating
+  data from legacy store to new one, and further optimizations and garbage collection
+
+Note that the store used in previous vesions (now called _legacy store_) does
+_not_ go away! Since the new store is not yet production ready, it remains the default
+store in Keycloak 19. Also other databases (MySQL, MariaDB, MS SQL Server, Oracle) are not
+supported by the new store, these are only supported by the legacy. We do consider
+second-level support for these databases, based on results of a community survey.
+
+For development and testing purposes, we offer a fast in-memory store called `chm`.
+This store is intentionally not capable of zero-downtime upgrade as that is not needed
+for development. As such, it offers also a quick Keycloak-without-database deployment.
+
+We will be glad to hear your feedback in https://github.com/keycloak/keycloak/discussions/new?category=keycloak-x-new-store&title=%5BUser%20feedback%5D%20&labels=team/storage-sig[GitHub Discussions].
+
+## Try it out
+
+The following section contains examples of how to configure Keycloak 19
+with the new store.
+
+WARNING: The command line options starting with `--storage` might change
+         without prior warning, based on the feedback from the community
+         and project needs.
+
+### See what Keycloak can do without a database
+
+After downloading, you can try Keycloak in a single-node deployment as simply as running
+
+   bin/kc.sh start-dev --storage=chm
+
+This starts Keycloak instance with a simple testing in-memory store which is persisted
+across Keycloak restarts but has no support for zero-downtime upgrades or clustering.
+
+### Next step: Persist data in relational database
+
+_Requirements:_ Postgres 14 or CockroachDB 22.1 (or newer)
+
+To keep the data in a database, run the following command:
+
+   bin/kc.sh start-dev --storage=jpa --db-url=<jdbc-url> --db-username=<username> --db-password=<password>
+
+Reference to https://github.com/keycloak/keycloak/discussions/9596[Q&A live document].
+
+CAUTION: This feature has known issues in CockroachDB, see https://github.com/keycloak/keycloak/issues/11618[this tracker for list].
+
+### Alternative next step: Persist data in external Infinispan
+
+_Requirements:_ Infinispan 12 (or newer)
+
+If you have a cluster of running Infinispan, you can use it for storing the data as well,
+just run the following command:
```

Comment:
Avoid "just" (as well as "simply") as those the reader's experience might vary.

```suggestion
and run the following command:
```

Reply:
rephrased